### PR TITLE
Add taskProgressBar and options

### PR DESCRIPTION
**Pull request by eschramm**
Closes https://github.com/CareEvolution/RKStudio-Participant/issues/789.

**ORK1Kit (RK1)**

For RK1, it was pretty straightforward, except that in order to have control over sizing the `UIProgressView`, I needed to embed into a container view. This is using the standard thickness to match platform UI (which will increase some in iOS 14).

|Text|Bar|None|
|---|---|---|
|1text|1bar|1none

**ResearchKit (RK2)**

In RK2, the standard behavior for displaying progress in text is to stuff it into a `UIBarButtonItem` due to moving the `title` into the navigation bar. Because of sizing, a `UIProgressView` is IMO too small to be usable there. Instead, this implementation places it into the RK2-specific navigation container which hovers from the bottom of the screen. Doing this requires some funky work to to connect the `ORKTaskViewController` to the `UIProgressView`. Due to order of operations in view life cycle, traversing the view hierarchy to find it doesn't work since at the point the block is fired, the view isn't entirely attached. Furthermore, because of updating the view after it appears, the changes to the `UIProgressView` animate strangely between steps. So ideally, when the `ORKNavigationContainerView` is created, it would know what the last step's state was as far as showing or hiding the `UIProgressView`. This necessitated a singleton to store that state to minimize non-sensical animation.

|Text|Bar|None|
|---|---|---|
|2text|2bar|2none

### CODE REVIEW

**Comment by mmertsock** [comment]
> This necessitated a singleton to store that state to minimize non-sensical animation.

Is this the first use of a singleton for UI behavior in ResearchKit?

File: ORK1Kit/ORK1Kit/Common/CEVRK1NavigationBarProgressView.m [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+
+@implementation CEVRK1NavigationBarProgressView {
+    UIProgressView *_progressView;
+}
+
+- (instancetype)initWithFrame:(CGRect)frame {
+    if (self = [super initWithFrame:frame]) {
+        [self createConstraints];
+        self.accessibilityLabel = @"NavigationBarProgressView";
```
Comment (mmertsock) [inline ORK1Kit/ORK1Kit/Common/CEVRK1NavigationBarProgressView.m]: Not a user-friendly string/needs localization

File: ORK1Kit/ORK1Kit/Common/CEVRK1NavigationBarProgressView.m [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        return self;
+    }
+    return nil;
+}
+
+- (void)createConstraints {
+    _progressView = [[UIProgressView alloc] initWithFrame:CGRectZero];
+    _progressView.progress = 0;
+    _progressView.accessibilityLabel = @"TaskProgressView";
```
Comment (mmertsock) [inline ORK1Kit/ORK1Kit/Common/CEVRK1NavigationBarProgressView.m]: Not a user-friendly string/needs localization

File: ORK1Kit/ORK1Kit/Common/CEVRK1NavigationBarProgressView.m [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    self.accessibilityElements = @[_progressView];
+    self.translatesAutoresizingMaskIntoConstraints = NO;
+    [self addSubview:_progressView];
+    _progressView.translatesAutoresizingMaskIntoConstraints = NO;
+    
+    // This forces the bar to stretch so the ORK1ProgressView will attempt to take up the entire available width
+    NSLayoutConstraint *widthConstraint = [self.widthAnchor constraintEqualToConstant:200];
+    widthConstraint.priority = NSURLSessionTaskPriorityHigh;
```
Comment (mmertsock) [inline ORK1Kit/ORK1Kit/Common/CEVRK1NavigationBarProgressView.m]: Use UILayoutPriorityDefaultHigh instead?

File: ORK1Kit/ORK1Kit/Common/ORK1TaskViewController.m [hunk 4]
```diff
@@ -966,11 +976,14 @@ - (void)showViewController:(ORK1StepViewController *)viewController goForward:(B
     // from the same VC.
     _currentStepViewController = viewController;
     
-    NSString *progressLabel = nil;
-    if ([self shouldDisplayProgressLabel]) {
+    ORK1TaskProgress taskProgress;
+    taskProgress.current = 0;
+    taskProgress.total =  0;
```
Comment (mmertsock) [inline ORK1Kit/ORK1Kit/Common/ORK1TaskViewController.m]: These 3 lines could be shortened to `ORK1TaskProgress taskProgress = ORK1TaskProgressMake(0, 0);`

File: ORK1Kit/ORK1Kit/Common/ORK1TaskViewController.m [hunk 5]
```diff
@@ -1491,7 +1538,7 @@ - (void)decodeRestorableStateWithCoder:(NSCoder *)coder {
         }
         
         if ([_task respondsToSelector:@selector(stepWithIdentifier:)]) {
-            _hasSetProgressLabel = [coder decodeBoolForKey:_ORK1HasSetProgressLabelRestoreKey];
+            _hasSetProgress = [coder decodeBoolForKey:_ORK1HasSetProgressRestoreKey];
```
Comment (mmertsock) [inline ORK1Kit/ORK1Kit/Common/ORK1TaskViewController.m]: Changing the restoration key may break the progress UI logic when restoring a task across ResearchKit version changes.

File: ResearchKit/Common/ORKNavigationContainerView.m [hunk 6]
```diff
@@ -263,6 +265,12 @@ - (void)setupParentStackView {
     _parentStackView.distribution = UIStackViewDistributionFill;
     
     [self addSubview:_parentStackView];
+    
+    CEVRKNavigationBarProgressView *progressView = [[CEVRKNavigationBarProgressView alloc] initWithFrame:CGRectZero];
+    self.taskProgressView = progressView;
+    [_parentStackView addArrangedSubview:progressView];
+    [progressView.topAnchor constraintEqualToAnchor:_parentStackView.topAnchor constant:10].active = YES;
+    [_parentStackView setCustomSpacing:20 afterView:progressView];
```
Comment (mmertsock) [inline ResearchKit/Common/ORKNavigationContainerView.m]: I didn't know about setCustomSpacing. Good to know!

File: ResearchKit/Common/ORKStepViewController.h [hunk 7]
```diff
@@ -357,6 +358,8 @@ ORK_CLASS_AVAILABLE
  */
 @property (nonatomic, weak, readonly, nullable) ORKTaskViewController *taskViewController;
 
+@property (nonatomic, weak, nullable) ORKNavigationContainerView *navigationContainerView;
```
Comment (mmertsock) [inline ResearchKit/Common/ORKStepViewController.h]: Would it follow ResearchKit idioms better to declare this in ORKStepViewController\_Internal.h?

**Review by mmertsock — COMMENTED** [review]
Code looks good overall. The core logic for progress calculation looks good to me, I thought through several use cases and it should be solid and match the web implementation.

File: ORK1Kit/ORK1Kit/Common/CEVRK1NavigationBarProgressView.m (hunk 1 above)
Comment (eschramm) [reply ORK1Kit/ORK1Kit/Common/CEVRK1NavigationBarProgressView.m]: I actually was able to remove these - they were remnants from creating the UITests. It turns out, putting views into the nav bar prevents them from being accessible ☹️

File: ORK1Kit/ORK1Kit/Common/CEVRK1NavigationBarProgressView.m (hunk 2 above)
Comment (eschramm) [reply ORK1Kit/ORK1Kit/Common/CEVRK1NavigationBarProgressView.m]: Same as above

File: ORK1Kit/ORK1Kit/Common/CEVRK1NavigationBarProgressView.m (hunk 3 above)
Comment (eschramm) [reply ORK1Kit/ORK1Kit/Common/CEVRK1NavigationBarProgressView.m]: Ah, good catch, thank you, really miss Swift for these things.

File: ORK1Kit/ORK1Kit/Common/ORK1TaskViewController.m (hunk 5 above)
Comment (eschramm) [reply ORK1Kit/ORK1Kit/Common/ORK1TaskViewController.m]: Good point, will revert.

File: ResearchKit/Common/ORKStepViewController.h (hunk 7 above)
Comment (eschramm) [reply ResearchKit/Common/ORKStepViewController.h]: Good suggestion, done.

**Comment by eschramm** [comment]
Was able to remove the singleton in RK2 by passing a reference to the StepViewController which can access the TaskViewController where we can store the last hidden state for the progress bar. 

I also found that simply embedding the `_parentStackView` inside another (`_grandparentStackView`) in the `ORKNavigationContainerView` would fix the wonky squishing/stretching seen in RK2 if we rotate to landscape mode.

BEFORE
Simulator Screen Shot - iPhone 11 - 2020-08-14 at 11 07 23


AFTER
Simulator Screen Shot - iPhone 11 - 2020-08-14 at 11 13 37

File: ResearchKit/Common/ORKStepViewController.h [hunk 8]
```diff
@@ -43,6 +43,7 @@ NS_ASSUME_NONNULL_BEGIN
 @class ORKStepResult;
 @class ORKStepViewController;
 @class ORKTaskViewController;
+@class ORKNavigationContainerView;
```
Comment (mmertsock) [inline ResearchKit/Common/ORKStepViewController.h]: This forward declaration should be moved to ORKSteViewController\_Internal.h

**Review by mmertsock — APPROVED** [review]
LGTM 🚢

**Pull request merged**